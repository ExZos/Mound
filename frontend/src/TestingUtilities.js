import GeneralComponent from './components/GeneralComponent';

class TestingUtilities extends GeneralComponent {
  addUserSessionItem(userID, spaceID, pollID=undefined) {
    const user = {
      id: userID,
      name: 'user' + userID,
      space: spaceID,
      space_name: 'space' + spaceID,
      space_status: true,
      poll: pollID
    };

    this.addToSessionArrayItem('users', user);
  }
}

export default TestingUtilities;
