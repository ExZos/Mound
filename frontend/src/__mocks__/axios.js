import { endpoints as api } from '../endpoints';

const axios = {
  create: () => axios,
  get: (url) => {
    switch(url) {
      case api.getSpaceByName + "Headon's Floor/":
        return Promise.resolve({
          data: {
            id: 1,
            name: "Headon's Floor",
            status: true
          }
        });
        break;

      default:
        return Promise.reject();
    }
  }
};

export default axios;
