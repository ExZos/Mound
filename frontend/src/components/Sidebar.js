import React from 'react';
import { Link } from 'react-router-dom';
import { Drawer } from '@material-ui/core';

import LandscapeIcon from '@material-ui/icons/Landscape';
import AddCircleOutlineIcon from '@material-ui/icons/AddCircleOutline';
import ClearAllIcon from '@material-ui/icons/ClearAll';

import GeneralComponent from './GeneralComponent';
import '../styles/sidebar.scss';

// TODO: new message indicators
class Sidebar extends GeneralComponent {
  componentDidMount(props) {
    this.sidebarContainer = document.getElementsByClassName('MuiDrawer-paper')[0];

    this.pushMainContent();
  }

  componentDidUpdate() {
    this.pushMainContent();
  }

  pushMainContent = () => {
    if(this.sidebarContainer) {
      document.body.style.marginLeft = this.sidebarContainer.offsetWidth + 'px';
    }
  }

  renderSpaces = () => {
    const users = this.getSessionItem('users');

    if(users) {
      return Object.values(users).map(user => (
        <div key={user.space} className="spaceNavItem ">
          <Link className={(user.space === this.props.spaceID) ? "active" : ""} tabIndex="-1" to={{
            pathname: "/r/",
            state: {
              pathname: "/s/",
              state: {
                space: {
                  id: user.space,
                  name: user.space_name,
                  status: user.space_status
                }
              }
            }
          }}>
            {user.space_name}
          </Link>
        </div>
      ));
    }
  }

  // TEMP clear link
  render() {
    return (
      <Drawer id="sidebarContainer" variant="permanent">
        <div id="sidebar">
          <div className="logo">
            <LandscapeIcon fontSize="large" />
          </div>

          <div className="spaceNav">
            {this.renderSpaces()}
          </div>

          <div className="menuItem">
            <Link to="/" tabIndex="-1">
              <AddCircleOutlineIcon fontSize="large" />
            </Link>
          </div>

          <div className="menuItem clear">
            <Link to="/" tabIndex="-1" onClick={this.clearSession}>
              <ClearAllIcon fontSize="large" />
            </Link>
          </div>
        </div>
      </Drawer>
    );
  }
}

export default Sidebar;
