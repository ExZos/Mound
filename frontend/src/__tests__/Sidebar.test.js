import React from 'react';
import { shallow, mount, render } from 'enzyme';
import axios from 'axios';
import { createBrowserHistory } from 'history';

import Sidebar from '../components/Sidebar';
import TestingUtilities from '../TestingUtilities';

const history = createBrowserHistory();
const tutils = new TestingUtilities();

describe('Sidebar', () => {
  beforeEach(() => {
    sessionStorage.clear();
  });

  it('should render correctly without props nor session', () => {
    const component = shallow(<Sidebar />);
    const tabs = component.find('div.spaceNavItem');

    expect(component).toMatchSnapshot();
    expect(tabs).toHaveLength(0);
  });

  it('should render tabs with session storage', () => {
    tutils.addUserSessionItem(1, 1, undefined);
    tutils.addUserSessionItem(2, 2, undefined);
    tutils.addUserSessionItem(3, 3, undefined);

    const component = shallow(<Sidebar />);
    const tabs = component.find('div.spaceNavItem');

    expect(component).toMatchSnapshot();
    expect(tabs).toHaveLength(3);
    expect(tabs.at(0).text()).toEqual('space1');
    expect(tabs.at(1).text()).toEqual('space2');
    expect(tabs.at(2).text()).toEqual('space3');
  });

  it('should highlight a tab with props and session', () => {
    tutils.addUserSessionItem(1, 1, undefined);
    tutils.addUserSessionItem(2, 2, undefined);
    tutils.addUserSessionItem(3, 3, undefined);

    const component = shallow(<Sidebar spaceID={2} />);
    const tab = component.find('Link.active');

    expect(tab).toHaveLength(1);
    expect(tab.children().text()).toEqual('space2');
  });

  it('should clear session on clear click', () => {
    tutils.addUserSessionItem(1, 1, undefined);
    tutils.addUserSessionItem(2, 2, undefined);
    tutils.addUserSessionItem(3, 3, undefined);

    const component = shallow(<Sidebar />);
    const clear = component.find('div.menuItem.clear Link');

    clear.props().onClick();

    expect(sessionStorage.getItem('users')).toBeNull();
  });
});
