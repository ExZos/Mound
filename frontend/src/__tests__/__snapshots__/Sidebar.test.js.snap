// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should render correctly without props nor session 1`] = `
<WithStyles(ForwardRef(Drawer))
  variant="permanent"
>
  <div
    id="sidebar"
  >
    <div
      className="logo"
    >
      <LandscapeIcon
        fontSize="large"
      />
    </div>
    <div
      className="spaceNav"
    />
    <div
      className="menuItem"
    >
      <Link
        tabIndex="-1"
        to="/"
      >
        <AddCircleOutlineIcon
          fontSize="large"
        />
      </Link>
    </div>
    <div
      className="menuItem clear"
    >
      <Link
        onClick={[Function]}
        tabIndex="-1"
        to="/"
      >
        <ClearAllIcon
          fontSize="large"
        />
      </Link>
    </div>
  </div>
</WithStyles(ForwardRef(Drawer))>
`;

exports[`Sidebar should render tabs with session storage 1`] = `
<WithStyles(ForwardRef(Drawer))
  variant="permanent"
>
  <div
    id="sidebar"
  >
    <div
      className="logo"
    >
      <LandscapeIcon
        fontSize="large"
      />
    </div>
    <div
      className="spaceNav"
    >
      <div
        className="spaceNavItem "
        key="1"
      >
        <Link
          className=""
          tabIndex="-1"
          to={
            Object {
              "pathname": "/r/",
              "state": Object {
                "pathname": "/s/",
                "state": Object {
                  "space": Object {
                    "id": 1,
                    "name": "space1",
                    "status": true,
                  },
                },
              },
            }
          }
        >
          space1
        </Link>
      </div>
      <div
        className="spaceNavItem "
        key="2"
      >
        <Link
          className=""
          tabIndex="-1"
          to={
            Object {
              "pathname": "/r/",
              "state": Object {
                "pathname": "/s/",
                "state": Object {
                  "space": Object {
                    "id": 2,
                    "name": "space2",
                    "status": true,
                  },
                },
              },
            }
          }
        >
          space2
        </Link>
      </div>
      <div
        className="spaceNavItem "
        key="3"
      >
        <Link
          className=""
          tabIndex="-1"
          to={
            Object {
              "pathname": "/r/",
              "state": Object {
                "pathname": "/s/",
                "state": Object {
                  "space": Object {
                    "id": 3,
                    "name": "space3",
                    "status": true,
                  },
                },
              },
            }
          }
        >
          space3
        </Link>
      </div>
    </div>
    <div
      className="menuItem"
    >
      <Link
        tabIndex="-1"
        to="/"
      >
        <AddCircleOutlineIcon
          fontSize="large"
        />
      </Link>
    </div>
    <div
      className="menuItem clear"
    >
      <Link
        onClick={[Function]}
        tabIndex="-1"
        to="/"
      >
        <ClearAllIcon
          fontSize="large"
        />
      </Link>
    </div>
  </div>
</WithStyles(ForwardRef(Drawer))>
`;
